pkg/server/endpoints/auth.go

```diff
@@ -0,0 +1,161 @@
+package endpoints
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/pkg/server/cache/dscache"
+	"github.com/spiffe/spire/pkg/server/svid"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+var (
+	misconfigLogMtx   sync.Mutex
+	misconfigLogTimes = make(map[spiffeid.TrustDomain]time.Time)
+	misconfigClk      = clock.New()
+)
+
+const misconfigLogEvery = time.Minute
+
+// shouldLogFederationMisconfiguration returns true if the last time a misconfiguration
+// was logged was more than misconfigLogEvery ago.
+func shouldLogFederationMisconfiguration(td spiffeid.TrustDomain) bool {
+	misconfigLogMtx.Lock()
+	defer misconfigLogMtx.Unlock()
+
+	now := misconfigClk.Now()
+	last, ok := misconfigLogTimes[td]
+	if !ok || now.Sub(last) >= misconfigLogEvery {
+		misconfigLogTimes[td] = now
+		return true
+	}
+	return false
+}
+
+// bundleGetter fetches the bundle for the given trust domain and parse it as x509 certificates.
+func (e *Endpoints) bundleGetter(ctx context.Context, td spiffeid.TrustDomain) ([]*x509.Certificate, error) {
+	commonServerBundle, err := e.DataStore.FetchBundle(dscache.WithCache(ctx), td.IDString())
+	if err != nil {
+		return nil, fmt.Errorf("get bundle from datastore: %w", err)
+	}
+	if commonServerBundle == nil {
+		if td != e.TrustDomain && shouldLogFederationMisconfiguration(td) {
+			e.Log.
+				WithField(telemetry.TrustDomain, td.String()).
+				Warn(
+					"No bundle found for foreign admin trust domain; admins from this trust domain will not be able to connect. " +
+						"Make sure this trust domain is correctly federated.",
+				)
+		}
+		return nil, fmt.Errorf("no bundle found for trust domain %q", td)
+	}
+
+	serverBundle, err := parseBundle(e.TrustDomain, commonServerBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	return serverBundle.X509Authorities(), nil
+}
+
+// serverSpiffeVerificationFunc returns a function that is used for peer certificate verification on TLS connections.
+// The returned function will verify that the peer certificate is valid, and apply a custom authorization with matchMemberOrOneOf.
+// If the peer certificate is not provided, the function will not make any verification and return nil.
+func (e *Endpoints) serverSpiffeVerificationFunc(bundleSource x509bundle.Source) func(_ [][]byte, _ [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		if rawCerts == nil {
+			return nil
+		}
+
+		return tlsconfig.VerifyPeerCertificate(
+			bundleSource,
+			tlsconfig.AdaptMatcher(matchMemberOrOneOf(e.TrustDomain, e.AdminIDs...)),
+		)(rawCerts, nil)
```

Comment:
nit: we should probably create this function before the closure to avoid unnecessary allocations on the handshake:
```suggestion
	verifyPeerCertificate := tlsconfig.VerifyPeerCertificate(
		bundleSource,
		tlsconfig.AdaptMatcher(matchMemberOrOneOf(e.TrustDomain, e.AdminIDs...)),
	)
	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
		if rawCerts == nil {
			return nil
		}
		return verifyPeerCertificate(rawCerts, nil)
```